## Importing the relevant libraries

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler,PolynomialFeatures
from sklearn.linear_model import LinearRegression
%matplotlib inline

## Loading the data

In [2]:
df = pd.read_csv("C:/Users/USER/Downloads/energydata_complete.csv")

In [3]:
df.head(2)

date  Appliances  lights     T1       RH_1    T2     RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.7900   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.7225   

      T3   RH_3    T4  ...         T9   RH_9     T_out  Press_mm_hg  RH_out  \
0  19.79  44.73  19.0  ...  17.033333  45.53  6.600000        733.5    92.0   
1  19.79  44.79  19.0  ...  17.066667  45.56  6.483333        733.6    92.0   

   Windspeed  Visibility  Tdewpoint        rv1        rv2  
0   7.000000   63.000000        5.3  13.275433  13.275433  
1   6.666667   59.166667        5.2  18.606195  18.606195  

[2 rows x 29 columns]

In [4]:
df.describe()

Appliances        lights            T1          RH_1            T2  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      97.694958      3.801875     21.686571     40.259739     20.341219   
std      102.524891      7.935988      1.606066      3.979299      2.192974   
min       10.000000      0.000000     16.790000     27.023333     16.100000   
25%       50.000000      0.000000     20.760000     37.333333     18.790000   
50%       60.000000      0.000000     21.600000     39.656667     20.000000   
75%      100.000000      0.000000     22.600000     43.066667     21.500000   
max     1080.000000     70.000000     26.260000     63.360000     29.856667   

               RH_2            T3          RH_3            T4          RH_4  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      40.420420     22.267611     39.242500     20.855335     39.026904   
std        4.069813      2.006111      3.254576      2.042884      4.341321   
min       20.463333     17.200000     28.766667     15.100000     27.660000   
25%       37.900000     20.790000     36.900000     19.530000     35.530000   
50%       40.500000     22.100000     38.530000     20.666667     38.400000   
75%       43.260000     23.290000     41.760000     22.100000     42.156667   
max       56.026667     29.236000     50.163333     26.200000     51.090000   

       ...            T9          RH_9         T_out   Press_mm_hg  \
count  ...  19735.000000  19735.000000  19735.000000  19735.000000   
mean   ...     19.485828     41.552401      7.411665    755.522602   
std    ...      2.014712      4.151497      5.317409      7.399441   
min    ...     14.890000     29.166667     -5.000000    729.300000   
25%    ...     18.000000     38.500000      3.666667    750.933333   
50%    ...     19.390000     40.900000      6.916667    756.100000   
75%    ...     20.600000     44.338095     10.408333    760.933333   
max    ...     24.500000     53.326667     26.100000    772.300000   

             RH_out     Windspeed    Visibility     Tdewpoint           rv1  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      79.750418      4.039752     38.330834      3.760707     24.988033   
std       14.901088      2.451221     11.794719      4.194648     14.496634   
min       24.000000      0.000000      1.000000     -6.600000      0.005322   
25%       70.333333      2.000000     29.000000      0.900000     12.497889   
50%       83.666667      3.666667     40.000000      3.433333     24.897653   
75%       91.666667      5.500000     40.000000      6.566667     37.583769   
max      100.000000     14.000000     66.000000     15.500000     49.996530   

                rv2  
count  19735.000000  
mean      24.988033  
std       14.496634  
min        0.005322  
25%       12.497889  
50%       24.897653  
75%       37.583769  
max       49.996530  

[8 rows x 28 columns]

## Question 12: From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two d.p.?

In [5]:
X = df[['T2']]
Y = df['T6']
lm = LinearRegression()
lm.fit(X,Y)
r2 = lm.score(X, Y)
round(r2,2)

0.64

## Answer: 0.64

## Question 13: Normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set. Answer the following questions:

## What is the Mean Absolute Error (in two decimal places)?

In [6]:
df.drop(columns = ['date', 'lights'], axis=1, inplace=True)

In [7]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
scaled_data = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)

In [8]:
X = scaled_data.drop(["Appliances"],axis=1)
y = scaled_data["Appliances"]

In [9]:
#Now, we split our dataset into the training and testing dataset.

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [10]:
reg = LinearRegression()
reg.fit(x_train,y_train)
y_pred = reg.predict(x_test)


In [11]:
#Mean Absolute Error(mae)
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(y_test,y_pred)
mae.round(2)


0.05

## Answer: 0.05

## Question 14: What is the Residual Sum of Squares (in two decimal places)?

In [12]:
import numpy as np
rss = np.sum(np.square(y_test - y_pred))
rss.round(2)

45.35

## Answer: 45.35

## Question 15: What is the Root Mean Squared Error (in three decimal places)?

In [13]:
from sklearn.metrics import mean_squared_error
mse = (mean_squared_error(y_test, y_pred))
rsme = np.sqrt(mse)
rsme.round(3)

0.088

## Answer: 0.088

## Question 16: What is the Coefficient of Determination (in two decimal places)? 

In [14]:
from sklearn.metrics import r2_score
r2_score = r2_score(y_test, y_pred)
r2_score.round(2)

0.15

## Answer: 0.15

## Question 17: Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

In [15]:
def weights(model,feat,col_name):
    weights = pd.Series(model.coef_,feat.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ["Features",col_name]
    weights_df[col_name].round(3)
    return weights_df
linear_reg_weights = weights(reg,x_train,"Linear_model_weight")
linear_reg_weights

Features  Linear_model_weight
0          RH_2            -0.456698
1         T_out            -0.321860
2            T2            -0.236178
3            T9            -0.189941
4          RH_8            -0.157595
5        RH_out            -0.077671
6          RH_7            -0.044614
7          RH_9            -0.039800
8            T5            -0.015657
9            T1            -0.003281
10          rv1             0.000770
11          rv2             0.000770
12  Press_mm_hg             0.006839
13           T7             0.010319
14   Visibility             0.012307
15         RH_5             0.016006
16         RH_4             0.026386
17           T4             0.028981
18    Windspeed             0.029183
19         RH_6             0.038049
20         RH_3             0.096048
21           T8             0.101995
22    Tdewpoint             0.117758
23           T6             0.236425
24           T3             0.290627
25         RH_1             0.553547

## Answer: RH_2 and RH_1

## Question 19: Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [16]:
from sklearn.linear_model import Ridge
ridge = Ridge(alpha=0.4)
ridge.fit(x_train,y_train)
ridge_pred = ridge.predict(x_test)

from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y_test,ridge_pred)
rmse = np.sqrt(mse)
rmse.round(3)

0.088

## Answer: No

## Question 18: Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [17]:
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(x_train, y_train)
lasso_pred = lasso_reg.predict(x_test)

In [18]:
#comparing the effects of regularisation


def get_weights_df(model, feat, col_name):
#this function returns the weight of every feature
    weights = pd.Series(model.coef_, feat.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ['Features', col_name]
    weights_df[col_name].round(3)
    return weights_df

In [19]:
linear_model_weights = get_weights_df(reg, x_train, 'Linear_Model_Weight')
ridge_weights_df = get_weights_df(ridge, x_train, 'Ridge_Weight')
lasso_weights_df = get_weights_df(lasso_reg, x_train, 'Lasso_weight')
final_weights = pd.merge(linear_model_weights, lasso_weights_df, on='Features')
final_weights

Features  Linear_Model_Weight  Lasso_weight
0          RH_2            -0.456698     -0.000000
1         T_out            -0.321860      0.000000
2            T2            -0.236178      0.000000
3            T9            -0.189941     -0.000000
4          RH_8            -0.157595     -0.000110
5        RH_out            -0.077671     -0.049557
6          RH_7            -0.044614     -0.000000
7          RH_9            -0.039800     -0.000000
8            T5            -0.015657     -0.000000
9            T1            -0.003281      0.000000
10          rv1             0.000770     -0.000000
11          rv2             0.000770     -0.000000
12  Press_mm_hg             0.006839     -0.000000
13           T7             0.010319     -0.000000
14   Visibility             0.012307      0.000000
15         RH_5             0.016006      0.000000
16         RH_4             0.026386      0.000000
17           T4             0.028981     -0.000000
18    Windspeed             0.029183      0.002912
19         RH_6             0.038049     -0.000000
20         RH_3             0.096048      0.000000
21           T8             0.101995      0.000000
22    Tdewpoint             0.117758      0.000000
23           T6             0.236425      0.000000
24           T3             0.290627      0.000000
25         RH_1             0.553547      0.017880

In [20]:
mse = mean_squared_error(y_test,lasso_pred)
rmse = np.sqrt(mse)
rmse.round(3)

0.094

## Answer: 0.094